### PR TITLE
Introduce DoctorValidatorsProvider to improve extensibility of flutter_tools

### PR DESCRIPTION
**Pull request by josephmr**
By injecting DoctorValidatorsProvider into the Doctor to provide validators, validators can be overridden without needing to override the whole Doctor. This makes the case of overriding validators less prone to complications should the Doctor API change.

### CODE REVIEW

File: packages/flutter_tools/lib/src/doctor.dart [hunk 1]
```diff
@@ -26,15 +26,20 @@ import 'vscode/vscode_validator.dart';
...
 
-class ValidatorTask {
-  ValidatorTask(this.validator, this.result);
-  final DoctorValidator validator;
-  final Future<ValidationResult> result;
+abstract class DoctorValidatorsProvider {
+  /// The singleton instance, pulled from the [AppContext].
+  static DoctorValidatorsProvider get instance => context[DoctorValidatorsProvider];
+
+  static final DoctorValidatorsProvider _defaultInstance = new _DefaultDoctorValidatorsProvider();
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/doctor.dart]: Just make this static final member public and avoid the getter

**Review by tvolkert — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/doctor.dart (hunk 1 above)
Comment (josephmr) [reply packages/flutter_tools/lib/src/doctor.dart]: Done.

**Comment by tvolkert** [comment]
Merging on red because of suspected flake.

**Pull request merged**